Collect test results into BuildKit-compatible JSON

tecosaur opened the pull request:
BuildKit can work with more granular test data, if we provide it in JSON form (https://buildkite.com/docs/test-analytics/importing-json).

This PR is inspired by Valentin's earlier attempt in #45799, which was pointed out to me in `#ci-dev` on Slack.

The hope is that by having more granular test data collected by the buildkite CI runs, it will be easier to see _which_ tests are making CI flaky, making the the path towards reliable CI a bit clearer.

Like #45799, tests results will need to be uploaded if they exist. See: https://github.com/JuliaCI/julia-buildkite/pull/150.

tecosaur committed:
Collect test results into BuildKit-compatible JSON

BuildKit can work with more granular test data, if we provide it in JSON
form (https://buildkite.com/docs/test-analytics/importing-json).

vchuravy commented on test/buildkitetestjson.jl:
```diff
@@ -0,0 +1,142 @@
+# Convert test(set) results to a Buildkit-compatible JSON representation.
+
+# Based on <https://buildkite.com/docs/test-analytics/importing-json#json-test-results-data-reference>
+# and Valentin Churavy's earlier work.
```
```suggestion
```

No need to acknowledge

vchuravy commented on test/runtests.jl:
```diff
@@ -423,6 +426,7 @@ cd(@__DIR__) do
         Test.push_testset(fake)
         Test.record(o_ts, fake)
         Test.pop_testset()
+        open(write_testset_json, "results.json", "w")
```
Should we add `results.json` to `.gitignore`?

tecosaur replied:
Probably. I've also spotted this is also not quite right. I'll make both changes now, and remove the attribution too.

tecosaur commented:
Spotted an issue with the `json_repr` for string, `escape_string(io, str)` doesn't escape quotes by default, so I've just fixed that. 

Arguably I should have done this earlier, but I just ran `./julia test/runtests.jl strings/annotated` to test this, and there was just an empty `results.json`. Adding a `@show o_ts.results` statement indicates that only non-passing tests are recorded.

To provide an example, I added a bogus failing test. Here's the JSON produced as a sample:

```json
[
  {
    "file_name": "/home/tec/Documents/Projects/julia/test/strings/annotated.jl",
    "history": {
      "duration": 3.5984139442443848,
      "start_at": 1.70680189927105e9,
      "end_at": 1.706801902869464e9
    },
    "name": "test: 1 == 1 + 1",
    "location": "/home/tec/Documents/Projects/julia/test/strings/annotated.jl:4",
    "failure_reason": "Evaluated: 1 == 2",
    "scope": "/Overall/strings/annotated",
    "id": "09650276-069c-32eb-54e3-16097ae71ad5",
    "result": "failed"
  }
]
```

vchuravy commented on test/runtests.jl:
```diff
@@ -424,6 +427,9 @@ cd(@__DIR__) do
         Test.pop_testset()
     end
+
+    open(io -> write_testset_json(io, o_ts), "results.json", "w")
```
Should we guard this by something like `Base.get_bool_env("CI", false)`

tecosaur replied:
I imagine this isn't much use to people running tests locally, I think we might as well. Pushed.

IanButterworth commented on test/buildkitetestjson.jl:
```diff
@@ -0,0 +1,142 @@
+end
+
+function result_dict(result::Test.Result)
+    file, line = if isnothing(result.source)
```
```suggestion
    file, line = if result isa Broken || isnothing(result.source) # Broken does not have a `source` field
```

tecosaur replied:
How about `if !hasproperty(result, :source) || isnothing(result.source)`?

IanButterworth commented on test/runtests.jl:
```diff
@@ -424,6 +427,10 @@ cd(@__DIR__) do
         Test.pop_testset()
     end
+
+    Base.get_bool_env("CI", false) &&
+        open(io -> write_testset_json(io, o_ts), "results.json", "w")
```
Probably good to give an indication it's been written
```suggestion
    if Base.get_bool_env("CI", false)
        open(io -> write_testset_json(io, o_ts), "results.json", "w")
        @info "CI detected. Test results written to `results.json` ($(Base.format_bytes(filesize("results.json"))))" 
    end
```

IanButterworth replied:
Given CI is green this could be for a future PR

vchuravy replied:
Maybe we should add that with some information to where we are writing the file since our upload step is still failing :)

vtjnash commented:
ERROR: LoadError: type Error has no field test_nonbool
Stacktrace:
  [1] getproperty
    @ ./Base.jl:49 [inlined]
  [2] add_failure_info!(data::Dict{String, Any}, result::Test.Error)

tecosaur commented:
Ah darn it, sorry about that typo that slipped in. I'll make a quickfix PR.
